src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java

```
@@ -168,49 +173,91 @@ void preprocessIngestDocument(IngestDocument ingestDocument) {
+                } catch (Exception e) {
+                    updateWithExceptions(ingestDocumentWrappers, e);
+                    handler.accept(ingestDocumentWrappers);
                 }
-                List<?> inferenceResults = results.subList(startIndex, startIndex + dataForInference.getInferenceList().size());
-                startIndex += dataForInference.getInferenceList().size();
-                setVectorFieldsToDocument(
-                    dataForInference.getIngestDocumentWrapper().getIngestDocument(),
-                    dataForInference.getProcessMap(),
-                    inferenceResults
-                );
-            }
-            handler.accept(ingestDocumentWrappers);
-        }, exception -> {
-            for (IngestDocumentWrapper ingestDocumentWrapper : ingestDocumentWrappers) {
-                // The IngestDocumentWrapper might already run into exception and not sent for inference. So here we only
-                // set exception to IngestDocumentWrapper which doesn't have exception before.
-                if (ingestDocumentWrapper.getException() == null) {
-                    ingestDocumentWrapper.update(ingestDocumentWrapper.getIngestDocument(), exception);
+            }, exception -> {
+                try {
+                    for (IngestDocumentWrapper ingestDocumentWrapper : ingestDocumentWrappers) {
+                        // The IngestDocumentWrapper might already run into exception and not sent for inference. So here we only
+                        // set exception to IngestDocumentWrapper which doesn't have exception before.
+                        if (ingestDocumentWrapper.getException() == null) {
+                            ingestDocumentWrapper.update(ingestDocumentWrapper.getIngestDocument(), exception);
+                        }
+                    }
+                    handler.accept(ingestDocumentWrappers);
+                } catch (Exception e) {
```

Why we need this try catch? If we can throw the exception when we update the exception in ingestDocumentWrappers then in the catch block we will throw exception again and will not invoke  `handler.accept`. If we can throw the exception when we do `handler.accept` shouldn't the exception handler of the handler to handle it rather than handle it here?

---

Good catch. `handler.accept()` should go to `finally` block.